lib/runtime/storage/storagediff.go

```diff
@@ -0,0 +1,308 @@
+	"golang.org/x/exp/slices"
+
+	"github.com/ChainSafe/gossamer/pkg/trie"
+)
+
+// storageDiff is a structure that stores the differences between consecutive
+// states of a trie, such as those occurring during the execution of a block.
+// It records updates (upserts), deletions, and changes to child tries.
+// This mechanism facilitates applying state transitions efficiently.
+// Changes accumulated in storageDiff can be applied to a trie using
+// the `applyToTrie` method
+type storageDiff struct {
+	upserts        map[string][]byte
+	deletes        map[string]bool
+	childChangeSet map[string]*storageDiff
+	mtx            sync.RWMutex
+}
+
+// newChangeSet initialises and returns a new storageDiff instance
+func newStorageDiff() *storageDiff {
+	return &storageDiff{
+		upserts:        make(map[string][]byte),
+		deletes:        make(map[string]bool),
+		childChangeSet: make(map[string]*storageDiff),
+		mtx:            sync.RWMutex{},
+	}
+}
+
+// get retrieves the value associated with the key if it's present in the
+// change set and returns a boolean indicating if the key is marked for deletion
+func (cs *storageDiff) get(key string) ([]byte, bool) {
+	if cs == nil {
+		return nil, false
+	}
+
+	cs.mtx.RLock()
+	defer cs.mtx.RUnlock()
+
+	// Check in recent upserts if not found check if we want to delete it
+	if val, ok := cs.upserts[key]; ok {
+		return val, false
+	} else if deleted := cs.deletes[key]; deleted {
+		return nil, true
+	}
+
+	return nil, false
+}
+
+// upsert records a new value for the key, or updates an existing value.
+// If the key was previously marked for deletion, that deletion is undone
+func (cs *storageDiff) upsert(key string, value []byte) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	// If we previously deleted this trie we have to undo that deletion
+	if cs.deletes[key] {
+		delete(cs.deletes, key)
+	}
+
+	cs.upserts[key] = value
+}
+
+// delete marks a key for deletion and removes it from upserts and
+// child changesets, if present.
+func (cs *storageDiff) delete(key string) {
+	if cs == nil {
+		return
+	}
+
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+
+	delete(cs.childChangeSet, key)
+	delete(cs.upserts, key)
+	cs.deletes[key] = true
+}
+
+// deleteChildLimit deletes lexicographical sorted keys from a child trie with
+// a maximum limit, potentially marking the entire child trie for deletion
+// if the limit is exceeded.
+// This method will do not consider keys created during actual block execution
+// https://spec.polkadot.network/chap-host-api#id-version-2-prototype-2
+func (cs *storageDiff) deleteChildLimit(keyToChild string,
+	currentChildKeys []string, limit int) (
+	deleted uint32, allDeleted bool) {
+	childChanges := cs.childChangeSet[keyToChild]
+	if childChanges == nil {
+		childChanges = newStorageDiff()
+	}
+
+	if limit == -1 {
+		cs.delete(keyToChild)
+		deletedKeys := len(childChanges.upserts) + len(currentChildKeys)
+		return uint32(deletedKeys), true
```

Comment:
> // This method will do not consider keys created during actual block execution

So should this method no be part of the `storageDiff` struct since this struct contains keys created under actual block execution?

Reply:
I might change the comment, they are not considered in the `limit` only

Reply:
Changed